pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -198,54 +198,54 @@ func (r *ReconcileBareMetalHost) Reconcile(request reconcile.Request) (result re
 		return result, err
 	}
 
-	// Check for a "discovered" host vs. one that we have all the info for.
-	if host.Spec.BMC.Address == "" {
-		reqLogger.Info(bmc.MissingAddressMsg)
-		dirty := host.SetOperationalStatus(metalkubev1alpha1.OperationalStatusDiscovered)
-		if dirty {
-			err = r.saveStatus(host)
-			if err != nil {
+	// Retrieve the BMC details from the host spec and validate host
+	// BMC details and build the credentials for talking to the
+	// management controller.
+	bmcCreds, bmcCredsSecret, err := r.buildAndValidateBMCCredentials(request, host)
+	if err != nil {
+		switch err.(type) {
+		// We key on empty bmc address and empty bmc credentials fields as a
+		// trigger the host needs to be put into a discovered status and requeue it for
+		// later.  We do set an error message (but not an error state) on the host
+		// so we can understand what we may be waiting on.
+		case *EmptyBMCAddressError, *EmptyBMCSecretError:
+			dirty := host.SetOperationalStatus(metalkubev1alpha1.OperationalStatusDiscovered)
+			if dirty {
+				// Set the host error message directly
+				// as we cannot use SetErrorCondition which
+				// overwrites our discovered state
+				host.Status.ErrorMessage = err.Error()
+				saveErr := r.saveStatus(host)
+				if saveErr != nil {
+					return reconcile.Result{Requeue: true}, saveErr
+				}
 				// Only publish the event if we do not have an error
 				// after saving so that we only publish one time.
 				r.publishEvent(request,
-					host.NewEvent("Discovered", "Discovered host without BMC address"))
+					host.NewEvent("Discovered", fmt.Sprintf("Discovered host with unusable BMC details: %s", err.Error())))
 			}
-			// Without the address we can't do any more so we return here
-			// without checking for an error.
-			return reconcile.Result{Requeue: true}, err
-		}
-		reqLogger.Info("nothing to do for discovered host without BMC address")
-		return reconcile.Result{}, nil
-	}
-	if host.Spec.BMC.CredentialsName == "" {
-		reqLogger.Info(bmc.MissingCredentialsMsg)
-		dirty := host.SetOperationalStatus(metalkubev1alpha1.OperationalStatusDiscovered)
-		if dirty {
-			err = r.saveStatus(host)
-			if err != nil {
-				// Only publish the event if we do not have an error
-				// after saving so that we only publish one time.
-				r.publishEvent(request,
-					host.NewEvent("Discovered", "Discovered host without BMC credentials"))
+			return reconcile.Result{Requeue: true, RequeueAfter: hostErrorRetryDelay}, nil
```

Comment:
There's no need to requeue here. Fixing an error related to missing BMC info will require editing the host, which will automatically trigger reconciliation again.

Reply:
Fixed